### PR TITLE
Deferred: Make jQuery.when synchronous when possible

### PR DESCRIPTION
**Pull request by gibson042**
Closes gh-3102
Fixes gh-3100

I'm ambivalent about this, but it doesn't cost much.

```
   raw     gz Sizes
264011  78513 dist/jquery.js
 86318  30014 dist/jquery.min.js

   raw     gz Compared to master @ 7f2ebd2c4dea186d7d981b939e6e2983a9d7f9c1
  +110    +22 dist/jquery.js
   +44    +17 dist/jquery.min.js
```

### CODE REVIEW

File: test/unit/deferred.js [hunk 1]
```diff
@@ -1049,3 +1049,24 @@ QUnit.test( "jQuery.when - notify does not affect resolved", function( assert )
...
+
+	assert.expect( 5 );
+
+	var when = "before",
+		resolved = jQuery.Deferred().resolve( true ),
+		rejected = jQuery.Deferred().reject( false ),
+		validate = function() {
+			assert.equal( when, "before" );
+		},
+		done = assert.async( 5 );
```
Comment (dmethvin) [inline test/unit/deferred.js]: I always forget you can do this.

**Comment by dmethvin** [comment]
I'm okay with it. It actually simplifies the docs and upgrade guide I think?

**Comment by gibson042** [comment]
I think so, just by virtue of changing less.

File: test/unit/deferred.js (hunk 1 above)
Comment (timmywil) [inline test/unit/deferred.js]: Didn't it require a QUnit update?

**Comment by timmywil** [comment]
I guess I'm okay with this.

File: test/unit/deferred.js (hunk 1 above)
Comment (gibson042) [inline test/unit/deferred.js]: We've only had it since January. It was added in QUnit 1.20.

**Comment by marcandre** [comment]
Great! Eager to run the spec suite of Parsley with this in. Thanks for the consideration and good discutions :smile:

**Pull request closed without merging**